node/worker/worker.go

```diff
@@ -41,32 +43,118 @@ type Worker struct {
+func (w *Worker) throttleTxs(selected types.Transactions, recentTxsStats types.RecentTxsStats, txsThrottleConfig *shardingconfig.TxsThrottleConfig, tx *types.Transaction) (common.Address, shardingconfig.TxThrottleFlag) {
+	chainID := tx.ChainID()
+	// Depending on the presence of the chain ID, sign with EIP155 or homestead
+	var s types.Signer
+	if chainID != nil {
+		s = types.NewEIP155Signer(chainID)
+	} else {
+		s = types.HomesteadSigner{}
+	}
+
+	var sender common.Address
+	msg, err := tx.AsMessage(s)
+	if err != nil {
+		utils.GetLogInstance().Error("Error when parsing tx into message",
+			"tx Id", tx.Hash().Hex(), "err", err)
+	} else {
+		sender = msg.From()
+	}
+
+	// already selected max num txs
+	if len(selected) > txsThrottleConfig.MaxNumTxsPerBlockLimit {
+		utils.GetLogInstance().Info("Throttling tx with max num txs per block limit",
+			"tx Id", tx.Hash().Hex(),
+			"MaxNumTxsPerBlockLimit", txsThrottleConfig.MaxNumTxsPerBlockLimit)
+		return sender, shardingconfig.TxUnselect
+	}
+
+	// throttle a single sender sending too many transactions in one block
+	if tx.Value().Cmp(txsThrottleConfig.MaxTxAmountLimit) > 0 {
+		utils.GetLogInstance().Info("Throttling tx with max amount limit",
+			"tx Id", tx.Hash().Hex(),
+			"MaxTxAmountLimit", txsThrottleConfig.MaxTxAmountLimit.Uint64(),
+			"Tx amount", tx.Value().Uint64())
+		return sender, shardingconfig.TxInvalid
+	}
+
+	// throttle too large transaction
+	var numTxsPastHour uint64
+	for _, blockTxsCounts := range recentTxsStats {
+		numTxsPastHour += blockTxsCounts[sender]
+	}
+	if numTxsPastHour >= txsThrottleConfig.MaxNumRecentTxsPerAccountLimit {
+		utils.GetLogInstance().Info("Throttling tx with max txs per account in a single block limit",
+			"tx Id", tx.Hash().Hex(),
+			"MaxNumRecentTxsPerAccountLimit", txsThrottleConfig.MaxNumRecentTxsPerAccountLimit)
+		return sender, shardingconfig.TxInvalid
+	}
+
+	return sender, shardingconfig.TxSelect
+}
+
 // SelectTransactionsForNewBlock selects transactions for new block.
-func (w *Worker) SelectTransactionsForNewBlock(txs types.Transactions, maxNumTxs int, coinbase common.Address) (types.Transactions, types.Transactions, types.Transactions) {
+func (w *Worker) SelectTransactionsForNewBlock(newBlockNum uint64, txs types.Transactions, recentTxsStats types.RecentTxsStats, txsThrottleConfig *shardingconfig.TxsThrottleConfig, coinbase common.Address) (types.Transactions, types.Transactions, types.Transactions) {
 	if w.current.gasPool == nil {
 		w.current.gasPool = new(core.GasPool).AddGas(w.current.header.GasLimit)
 	}
+
 	selected := types.Transactions{}
 	unselected := types.Transactions{}
 	invalid := types.Transactions{}
 	for _, tx := range txs {
 		if tx.ShardID() != w.shardID {
 			invalid = append(invalid, tx)
+			continue
 		}
-		snap := w.current.state.Snapshot()
-		_, err := w.commitTransaction(tx, coinbase)
-		if len(selected) > maxNumTxs {
+
+		sender, flag := w.throttleTxs(selected, recentTxsStats, txsThrottleConfig, tx)
```

Comment:
moving throttling logic to addPendingTxs will make the logic cleaner. see https://github.com/harmony-one/harmony/pull/1310

Reply:
Agreed. Let me make that as a follow up.